### ISSUE TITLE
Pagination for Users/Groups

### ISSUE THREAD

**Issue by andriiunitq** [okta/okta-sdk-java#562]
### I'm submitting a
  - [ ] feature request

### Background info

I want to build pagination for Users/Groups on the UI, so in other words, I want to get any page, based on the page number.
However, according to my understanding, the library provides wrappers i.e. users/groups - UserList, GroupList and these classes under the hood are using a custom implementation of Iterator ( called Page iterator), so they automatically obtain the next page based on the "Link" header during iteration.

So, to be able to get data from a particular page I need to use something like this:
`final int SKIP_ELEMENTS = PAGE_LIMIT * PAGE_NUMBER;

final List<Group> groupList = client.getDataStore()
     .getResource("/api/v1/groups?limit=" + PAGE_LIMIT, GroupList.class)
    .stream()
    .skip(SKIP_ELEMENTS)
    .limit(PAGE_LIMIT)
    .collect(Collectors.toList());`

So, my concern that to be able to get for example the third page, under the hood page iterator will obtain 2 previous pages and after that obtains the third. Am I right? and again when I try to obtain the fourth page, the iterating will start from scratch (from the first page)?

### Expected behavior
to avoid obtain redundant elements (pages) and consider extending API to use page number.

**Comment by arvindkrishnakumar-okta** [comment]
@andriiunitq Thanks for your question! 

Pagination is not based on 'page number' but is tied to cursors that are opaque to the caller (Ref: https://developer.okta.com/docs/reference/api-overview/#pagination).

A potential way to go about your use case is to follow the URLs in `LinkHeader` (store the cursors) and use them in the search query parameters `before`, `after` and `limit` to fetch the 'page' you want.

Let me know if that helps!

**Comment by andriiunitq** [comment]
@arvindkrishnakumar-okta yes, I need to use LinkHeader directly and it also means that to be able to obtain for example the third page and I need to obtain 2 cursors ( for the 1st and 2nd pages) if I understood correctly?

**Comment by bdemers** [comment]
Answering mobile, so forgive any errors.

You should be able to manually page using this technique:

https://github.com/okta/okta-sdk-java#call-other-api-endpoints

You WILL need to “.limit()” the stream to prevent retrieval of the next page.

Then call ‘.getProperty(“nextPage”)’ on the collection resource.

This may require some casting (if so, that is something we could improve)

Keep us posted

-Brian

**Comment by andriiunitq** [comment]
Hi @bdemers 

yes thank your help and support, it works as you described and it requires some casting.

**Comment by nmarshall23** [comment]
@andriiunitq would you mind sharing your code? I've been trying to do something simpler on the` /api/v1/users` endpoint. 

The default paging behavior of `UserList` takes too long to load.

**Comment by bdemers** [comment]
Hey @nmarshall23 

This is untested, but It _should_ provide more detail than my previous [mobile] message.

You will need to use the "Call other endpoints" technique to request each page.

So for the user endpoint, it _should_ look something like this:


```java

int pageSize = 10;

// For details on pagination params take a look at:
// https://developer.okta.com/docs/reference/api-overview/#pagination

UserList usersPage1 = client.http()
    .addQueryParameter("limit", pageSize)
    .get("/api/v1/users", UserList.class);
```

This would get you the first page, but when you iterate over the collection you would need to limit the stream, for example:

```java
usersPage1.stream().limit(pageSize).forEach(user -> {
    log.debug("do something here", user);
})
```

To get the next page you would need to make a similar request as before, except change the to the "next page" URL (as @arvindkrishnakumar-okta mentioned it's cursored, so you would need to remember this URL between page loads (if that applies to you)

Getting the URL _should_ look something like this:

```java
String nextPageUrl = usersPage1.getProperty("nextPage");

// now make another request using this URL:
UserList usersPage2 = client.http()
    .addQueryParameter("limit", pageSize)
    .get(nextPageUrl, UserList.class);
```

**NOTE:** you may need to cast `usersPage1` to `com.okta.sdk.impl.resource.AbstractCollectionResource`
This means you would need to change the scope of your dependency in your `pom.xml` (or gradle build) to a `compile` scope.  We generally recommend against this, but assuming all of this works, we can make this more friendly in a future release (e.g. removing the cast and/or adding a `getNextPageUrl()` method.)


Once again, this is mostly from memory, and a little copy & pasting, so I didn't test this out, please report back to us! If it works, we can move forward with a nicer fix for a future version, and if it doesn't we will get you an actually tested code snippet 😄

**Comment by andriiunitq** [comment]
Hi @bdemers 
Thank you for your help and support, the obtaining the Users/Groups worked as you described.
`   final GroupList groupList = oktaClient.http()
        .setQueryParameters(getQueryParams(pageable))
        .get(apiPath, GroupList.class);`

However, for obtain the next page, I've used casting to another class
`(String) ((DefaultGroupList) groupList).getProperty("nextPage");`

**Comment by bdemers** [comment]
@arvindkrishnakumar-okta can you stick this on the backlog?  I think exposing something like a `getNextPageUrl()` method  in the `CollectionResource` interface should do the trick. Then manual paging should look something like:

```java
GroupList groupList = client.listGroups();

String nextPage = groupList.getNextPageUrl();

// iterate
groupList.stream()
    .limit(<page-size>)
    .forEach(group -> ... );

...

// manually fetch next page
GroupList groupList2 = client.http().get(nextPage, GroupList.class);
```

**Comment by arvindkrishnakumar-okta** [comment]
@bdemers sure will do.

**Comment by arvindkrishnakumar-okta** [comment]
Internal Ref: OKTA-382495
